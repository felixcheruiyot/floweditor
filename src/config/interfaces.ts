import { AnyAction } from '~/flowTypes';

export enum FlowTypes {
    MESSAGE = 'messaging',
    VOICE = 'voice',
    SURVEY = 'messaging_offline',
    NONE = '-'
}

export const enum Types {
    execute_actions = 'execute_actions',
    add_contact_urn = 'add_contact_urn',
    add_contact_groups = 'add_contact_groups',
    add_input_labels = 'add_input_labels',
    remove_contact_groups = 'remove_contact_groups',
    set_contact_channel = 'set_contact_channel',
    set_contact_field = 'set_contact_field',
    set_contact_name = 'set_contact_name',
    set_contact_language = 'set_contact_language',
    set_run_result = 'set_run_result',
    call_resthook = 'call_resthook',
    call_webhook = 'call_webhook',
    send_msg = 'send_msg',
    send_email = 'send_email',
    send_broadcast = 'send_broadcast',
    enter_flow = 'enter_flow',
    start_session = 'start_session',
    transfer_airtime = 'transfer_airtime',
    split_by_airtime = 'split_by_airtime',
    split_by_expression = 'split_by_expression',
    split_by_contact_field = 'split_by_contact_field',
    split_by_run_result = 'split_by_run_result',
    split_by_run_result_delimited = 'split_by_run_result_delimited',
    split_by_groups = 'split_by_groups',
    split_by_random = 'split_by_random',
    split_by_resthook = 'split_by_resthook',
    split_by_subflow = 'split_by_subflow',
    split_by_webhook = 'split_by_webhook',
    wait_for_response = 'wait_for_response',
    wait_for_menu = 'wait_for_menu',
    wait_for_digits = 'wait_for_digits',
    wait_for_audio = 'wait_for_audio',
    wait_for_video = 'wait_for_video',
    wait_for_location = 'wait_for_location',
    wait_for_image = 'wait_for_image',
    missing = 'missing',
    say_msg = 'say_msg',
    play_audio = 'play_audio'
}

export enum Operators {
    has_any_word = 'has_any_word',
    has_all_words = 'has_all_words',
    has_phrase = 'has_phrase',
    has_only_phrase = 'has_only_phrase',
    has_beginning = 'has_beginning',
    has_text = 'has_text',
    has_date = 'has_date',
    has_date_lt = 'has_date_lt',
    has_date_eq = 'has_date_eq',
    has_date_gt = 'has_date_gt',
    has_number = 'has_number',
    has_number_between = 'has_number_between',
    has_number_lt = 'has_number_lt',
    has_number_lte = 'has_number_lte',
    has_number_eq = 'has_number_eq',
    has_number_gte = 'has_number_gte',
    has_number_gt = 'has_number_gt',
    has_run_status = 'has_run_status',
    is_text_eq = 'is_text_eq',
    has_group = 'has_group',
    has_phone = 'has_phone',
    has_email = 'has_email',
    has_error = 'has_error',
    has_value = 'has_value',
    has_attachment = 'has_attachment',
    has_wait_timed_out = 'has_wait_timed_out',
    has_district = 'has_district',
    has_state = 'has_state',
    has_ward = 'has_ward',
    has_pattern = 'has_pattern',
    has_webhook_status = 'has_webhook_status',
    is_error = 'is_error'
}

export interface FlowTypeVisibility {
    visibility?: FlowTypes[];
}

export interface Type extends FlowTypeVisibility {
    type: Types;
    name: string;
    description: string;
    component?: React.SFC<AnyAction>;
    form?: React.ComponentClass<any>;
    aliases?: string[];
    localization?: React.ComponentClass<any>;
    localizeableKeys?: string[];
}

export interface Operator extends FlowTypeVisibility {
    type: Operators;
    verboseName: string;
    additionalArguments: number;
    initialArgument?: string;
    categoryName?: string;
}

export interface OperatorMap {
    [propName: string]: Operator;
}

export const HIDDEN = [FlowTypes.NONE];
export const VOICE = [FlowTypes.VOICE];
export const SURVEY = [FlowTypes.SURVEY];
export const TEXT_TYPES = [FlowTypes.MESSAGE, FlowTypes.SURVEY];
export const ONLINE = [FlowTypes.MESSAGE, FlowTypes.VOICE];
