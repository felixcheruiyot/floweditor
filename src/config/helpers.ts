import { FlowTypes, FlowTypeVisibility, Operator, Type } from '~/config/interfaces';

export const filterOperators = (operators: Operator[], flowType: FlowTypes): Operator[] => {
    return filterVisibility(operators, flowType);
};

export const filterTypeConfigs = (typeConfigs: Type[], flowType: FlowTypes): Type[] => {
    return filterVisibility(typeConfigs, flowType);
};

const filterVisibility = (items: FlowTypeVisibility[], flowType: FlowTypes): any[] => {
    return items.filter((item: FlowTypeVisibility) => {
        if (item.visibility === undefined) {
            return true;
        }

        return item.visibility.findIndex((ft: FlowTypes) => ft === flowType) > -1;
    });
};
