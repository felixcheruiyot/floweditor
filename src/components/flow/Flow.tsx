import { react as bindCallbacks } from 'auto-bind';
import * as React from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { Canvas } from '~/components/canvas/Canvas';
import { CanvasDraggableProps } from '~/components/canvas/CanvasDraggable';
import Node from '~/components/flow/node/Node';
import { getDraggedFrom } from '~/components/helpers';
import NodeEditor from '~/components/nodeeditor/NodeEditor';
import Simulator from '~/components/simulator/Simulator';
import Sticky, { STICKY_BODY, STICKY_TITLE } from '~/components/sticky/Sticky';
import { ConfigProviderContext, fakePropType } from '~/config/ConfigProvider';
import { Exit, FlowDefinition } from '~/flowTypes';
import Plumber from '~/services/Plumber';
import editorState, { DragSelection, EditorState } from '~/store/editor';
import { RenderNode } from '~/store/flowContext';
import { detectLoops, getOrderedNodes } from '~/store/helpers';
import { NodeEditorSettings } from '~/store/nodeEditor';
import AppState from '~/store/state';
import {
    ConnectionEvent,
    DispatchWithState,
    ensureStartNode,
    EnsureStartNode,
    mergeEditorState,
    MergeEditorState,
    NoParamsAC,
    OnConnectionDrag,
    onConnectionDrag,
    OnOpenNodeEditor,
    onOpenNodeEditor,
    OnUpdateCanvasPositions,
    onUpdateCanvasPositions,
    resetNodeEditingState,
    UpdateConnection,
    updateConnection,
    updateSticky,
    UpdateSticky
} from '~/store/thunks';
import { contextTypes } from '~/testUtils';
import {
    ACTIVITY_INTERVAL,
    createUUID,
    isRealValue,
    NODE_PADDING,
    renderIf,
    snapToGrid,
    timeEnd,
    timeStart
} from '~/utils';
import Debug from '~/utils/debug';

export interface FlowStoreProps {
    editorState: Partial<EditorState>;
    mergeEditorState: MergeEditorState;

    definition: FlowDefinition;
    nodes: { [uuid: string]: RenderNode };
    dependencies: FlowDefinition[];
    nodeEditorSettings: NodeEditorSettings;

    ensureStartNode: EnsureStartNode;
    updateConnection: UpdateConnection;
    onOpenNodeEditor: OnOpenNodeEditor;
    onUpdateCanvasPositions: OnUpdateCanvasPositions;
    resetNodeEditingState: NoParamsAC;
    onConnectionDrag: OnConnectionDrag;
    updateSticky: UpdateSticky;
}

export interface Translations {
    [uuid: string]: any;
}

export const DRAG_THRESHOLD = 3;
export const REPAINT_TIMEOUT = 500;
export const GHOST_POSITION_INITIAL = { left: -1000, top: -1000 };

export const nodeSpecId = 'node';
export const nodesContainerSpecId = 'node-container';
export const ghostNodeSpecId = 'ghost-node';
export const dragSelectSpecId = 'drag-select';

export const isDraggingBack = (event: ConnectionEvent) => {
    return event.suspendedElementId === event.targetId && event.source !== null;
};

export const getDragStyle = (drag: DragSelection) => {
    const left = Math.min(drag.startX, drag.currentX);
    const top = Math.min(drag.startY, drag.currentY);
    const width = Math.max(drag.startX, drag.currentX) - left;
    const height = Math.max(drag.startY, drag.currentY) - top;
    return {
        left,
        top,
        width,
        height
    };
};

export class Flow extends React.Component<FlowStoreProps, {}> {
    private Plumber: Plumber;
    private ele: HTMLDivElement;
    private nodeContainerUUID: string;

    // Refs
    private ghost: any;

    public static contextTypes = {
        config: fakePropType
    };

    constructor(props: FlowStoreProps, context: ConfigProviderContext) {
        super(props, context);

        this.nodeContainerUUID = createUUID();

        this.Plumber = new Plumber();

        /* istanbul ignore next */
        if (context.config.debug) {
            window.fe = new Debug(props, this.props.editorState.debug);
        }

        bindCallbacks(this, {
            include: [/Ref$/, /^on/, /^is/, /^get/]
        });

        timeStart('RenderAndPlumb');
    }

    private onRef(ref: HTMLDivElement): HTMLDivElement {
        return (this.ele = ref);
    }

    private ghostRef(ref: any): any {
        return (this.ghost = ref);
    }

    public componentDidMount(): void {
        this.Plumber.bind('connection', (event: ConnectionEvent) =>
            this.props.updateConnection(event.sourceId, event.targetId)
        );
        this.Plumber.bind('beforeDrag', (event: ConnectionEvent) =>
            this.beforeConnectionDrag(event)
        );

        this.Plumber.bind('connectionDrag', (event: ConnectionEvent) => {
            this.props.onConnectionDrag(event, this.context.config.flowType);
        });

        this.Plumber.bind('connectionDragStop', (event: ConnectionEvent) =>
            this.onConnectorDrop(event)
        );
        this.Plumber.bind(
            'beforeStartDetach',
            (event: ConnectionEvent) => !this.props.editorState.translating
        );
        this.Plumber.bind('beforeDetach', (event: ConnectionEvent) => true);
        this.Plumber.bind('beforeDrop', (event: ConnectionEvent) =>
            this.onBeforeConnectorDrop(event)
        );

        // If we don't have any nodes, create our first one
        this.props.ensureStartNode();

        let offset = { left: 0, top: 0 };

        /* istanbul ignore next */
        if (this.ele) {
            offset = this.ele.getBoundingClientRect();
        }

        this.props.mergeEditorState({
            containerOffset: { left: offset.left, top: offset.top + window.scrollY }
        });

        timeEnd('RenderAndPlumb');

        // deals with safari load rendering throwing
        // off the jsplumb offsets
        window.setTimeout(() => this.Plumber.repaint(), REPAINT_TIMEOUT);
    }

    public componentWillUnmount(): void {
        this.Plumber.reset();
    }

    public componentWillUpdate(prevProps: FlowStoreProps): void {
        if (
            prevProps.editorState.activityInterval === this.props.editorState.activityInterval &&
            this.props.editorState.activityInterval !== ACTIVITY_INTERVAL
        ) {
            this.props.mergeEditorState({ activityInterval: ACTIVITY_INTERVAL });
        }
    }

    /**
     * Called right before a connector is dropped onto a new node
     */
    private onBeforeConnectorDrop(event: ConnectionEvent): boolean {
        this.props.resetNodeEditingState();
        const fromNodeUUID = event.sourceId.split(':')[0];
        try {
            detectLoops(this.props.nodes, fromNodeUUID, event.targetId);
        } catch {
            return false;
        }
        return true;
    }

    /**
     * Called the moment a connector is done dragging, whether it is dropped on an
     * existing node or on to empty space.
     */
    private onConnectorDrop(event: ConnectionEvent): boolean {
        const { ghostNode } = this.props.editorState;
        // Don't show the node editor if we a dragging back to where we were
        if (isRealValue(ghostNode) && !isDraggingBack(event)) {
            // Wire up the drag from to our ghost node
            this.Plumber.recalculate(ghostNode.node.uuid);

            const dragPoint = getDraggedFrom(ghostNode);
            this.Plumber.connect(
                dragPoint.nodeUUID + ':' + dragPoint.exitUUID,
                ghostNode.node.uuid
            );

            // Save our position for later
            const { left, top } = (this.ghost &&
                snapToGrid(
                    this.ghost.wrappedInstance.ele.offsetLeft,
                    this.ghost.wrappedInstance.ele.offsetTop
                )) || { left: 0, top: 0 };

            this.props.editorState.ghostNode.ui.position = { left, top };
            let originalAction = null;
            if (ghostNode.node.actions && ghostNode.node.actions.length === 1) {
                originalAction = ghostNode.node.actions[0];
            }

            // Bring up the node editor
            this.props.onOpenNodeEditor({
                originalNode: ghostNode,
                originalAction
            });
        }

        /* istanbul ignore next */
        document.removeEventListener('mousemove', (window as any).ghostListener);

        return true;
    }

    private beforeConnectionDrag(event: ConnectionEvent): boolean {
        return !this.props.editorState.translating;
    }

    private getNodes(): CanvasDraggableProps[] {
        return getOrderedNodes(this.props.nodes).map((renderNode: RenderNode, idx: number) => {
            return {
                uuid: renderNode.node.uuid,
                position: renderNode.ui.position,
                ele: (selected: boolean) => (
                    <Node
                        startingNode={idx === 0}
                        selected={selected}
                        key={renderNode.node.uuid}
                        data-spec={nodeSpecId}
                        nodeUUID={renderNode.node.uuid}
                        plumberMakeTarget={this.Plumber.makeTarget}
                        plumberRemove={this.Plumber.remove}
                        plumberRecalculate={this.Plumber.recalculate}
                        plumberMakeSource={this.Plumber.makeSource}
                        plumberConnectExit={this.Plumber.connectExit}
                        plumberUpdateClass={this.Plumber.updateClass}
                    />
                )
            };
        });
    }

    private getStickies(): CanvasDraggableProps[] {
        const stickyMap = this.props.definition._ui.stickies || {};
        return Object.keys(stickyMap).map(uuid => {
            return {
                uuid,
                ele: (selected: boolean) => (
                    <Sticky key={uuid} uuid={uuid} sticky={stickyMap[uuid]} selected={selected} />
                ),
                position: stickyMap[uuid].position
            };
        });
    }

    private getDragNode(): JSX.Element {
        return isRealValue(this.props.editorState.ghostNode) ? (
            <div style={{ position: 'absolute', display: 'block' }}>
                <Node
                    selected={false}
                    startingNode={false}
                    ref={this.ghostRef}
                    key={this.props.editorState.ghostNode.node.uuid}
                    data-spec={ghostNodeSpecId}
                    ghost={true}
                    nodeUUID={this.props.editorState.ghostNode.node.uuid}
                    plumberMakeTarget={this.Plumber.makeTarget}
                    plumberRemove={this.Plumber.remove}
                    plumberRecalculate={this.Plumber.recalculate}
                    plumberMakeSource={this.Plumber.makeSource}
                    plumberConnectExit={this.Plumber.connectExit}
                    plumberUpdateClass={this.Plumber.updateClass}
                />
            </div>
        ) : null;
    }

    private getSimulator(): JSX.Element {
        return renderIf(
            this.context.config.endpoints && this.context.config.endpoints.simulateStart
        )(<Simulator mergeEditorState={this.props.mergeEditorState} />);
    }

    private getNodeEditor(): JSX.Element {
        return renderIf(this.props.nodeEditorSettings !== null)(
            <NodeEditor
                plumberConnectExit={this.Plumber.connectExit}
                plumberRepaintForDuration={this.Plumber.repaintForDuration}
            />
        );
    }

    private isClickOnCanvas(event: React.MouseEvent<HTMLDivElement>): boolean {
        // TODO: not sure the TS-safe way to access id here
        return (event.target as any).id === this.nodeContainerUUID;
    }

    private onDoubleClick(event: React.MouseEvent<HTMLDivElement>): void {
        if (this.isClickOnCanvas(event)) {
            const { left, top } = snapToGrid(
                event.pageX - this.props.editorState.containerOffset.left - 100 + NODE_PADDING,
                event.pageY - this.props.editorState.containerOffset.top - NODE_PADDING * 2 - 40
            );

            this.props.updateSticky(createUUID(), {
                position: { left, top },
                title: STICKY_TITLE,
                body: STICKY_BODY
            });
        }
    }

    public render(): JSX.Element {
        const draggables = this.getStickies().concat(this.getNodes());

        return (
            <div onDoubleClick={this.onDoubleClick} ref={this.onRef}>
                <Canvas
                    onDragging={(uuids: string[]) => {
                        uuids.forEach((uuid: string) => {
                            if (uuid in this.props.nodes) {
                                this.props.nodes[uuid].node.exits.forEach((exit: Exit) => {
                                    if (exit.destination_uuid) {
                                        uuids.push(uuid + ':' + exit.uuid);
                                    }
                                });
                            }
                        });
                        this.Plumber.recalculateUUIDs(uuids);
                    }}
                    uuid={this.nodeContainerUUID}
                    dragActive={this.props.editorState.dragActive}
                    mergeEditorState={this.props.mergeEditorState}
                    draggables={draggables}
                    onUpdatePositions={this.props.onUpdateCanvasPositions}
                >
                    {this.getSimulator()}
                    {this.getDragNode()}
                    {this.getNodeEditor()}
                </Canvas>
            </div>
        );
    }
}

/* istanbul ignore next */
const mapStateToProps = ({
    flowContext: { definition, dependencies, nodes },
    editorState,
    nodeEditor: { settings }
}: AppState) => {
    return {
        nodeEditorSettings: settings,
        definition,
        nodes,
        dependencies,
        editorState: editorState as Partial<EditorState>
    };
};

/* istanbul ignore next */
const mapDispatchToProps = (dispatch: DispatchWithState) =>
    bindActionCreators(
        {
            mergeEditorState,
            ensureStartNode,
            resetNodeEditingState,
            onConnectionDrag,
            onOpenNodeEditor,
            onUpdateCanvasPositions,
            updateConnection,
            updateSticky
        },
        dispatch
    );

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(Flow);
