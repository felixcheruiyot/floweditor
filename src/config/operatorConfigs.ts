import { HIDDEN, Operator, OperatorMap, Operators, TEXT_TYPES } from '~/config/interfaces';

const ARGUMENT_ATTACHMENTS = '@input.attachments';

export const operatorConfigList: Operator[] = [
    {
        type: Operators.has_any_word,
        verboseName: 'has any of the words',
        additionalArguments: 1,
        visibility: TEXT_TYPES
    },
    {
        type: Operators.has_all_words,
        verboseName: 'has all of the words',
        additionalArguments: 1,
        visibility: TEXT_TYPES
    },
    {
        type: Operators.has_phrase,
        verboseName: 'has the phrase',
        additionalArguments: 1,
        visibility: TEXT_TYPES
    },
    {
        type: Operators.has_only_phrase,
        verboseName: 'has only the phrase',
        additionalArguments: 1,
        visibility: TEXT_TYPES
    },
    {
        type: Operators.has_beginning,
        verboseName: 'starts with',
        additionalArguments: 1
    },
    {
        type: Operators.has_text,
        verboseName: 'has some text',
        additionalArguments: 0,
        categoryName: 'Has Text',
        visibility: TEXT_TYPES
    },
    {
        type: Operators.has_number,
        verboseName: 'has a number',
        additionalArguments: 0,

        categoryName: 'Has Number'
    },
    {
        type: Operators.has_number_between,
        verboseName: 'has a number between',
        additionalArguments: 2
    },
    {
        type: Operators.has_number_lt,
        verboseName: 'has a number below',
        additionalArguments: 1
    },
    {
        type: Operators.has_number_lte,
        verboseName: 'has a number at or below',
        additionalArguments: 1
    },
    {
        type: Operators.has_number_eq,
        verboseName: 'has a number equal to',
        additionalArguments: 1
    },
    {
        type: Operators.has_number_gte,
        verboseName: 'has a number at or above',
        additionalArguments: 1
    },
    {
        type: Operators.has_number_gt,
        verboseName: 'has a number above',
        additionalArguments: 1
    },
    {
        type: Operators.has_date,
        verboseName: 'has a date',
        additionalArguments: 0,

        categoryName: 'Has Date',
        visibility: TEXT_TYPES
    },
    {
        type: Operators.has_date_lt,
        verboseName: 'has a date before',
        additionalArguments: 1,

        visibility: TEXT_TYPES
    },
    {
        type: Operators.has_date_eq,
        verboseName: 'has a date equal to',
        additionalArguments: 1,

        visibility: TEXT_TYPES
    },
    {
        type: Operators.has_date_gt,
        verboseName: 'has a date after',
        additionalArguments: 1,

        visibility: TEXT_TYPES
    },
    {
        type: Operators.has_run_status,
        verboseName: 'has a run status of',
        additionalArguments: 1,
        visibility: HIDDEN
    },
    {
        type: Operators.has_group,
        verboseName: 'is in the group',
        additionalArguments: 1,
        visibility: HIDDEN
    },
    {
        type: Operators.has_phone,
        verboseName: 'has a phone number',
        additionalArguments: 0,
        categoryName: 'Has Phone'
    },
    {
        type: Operators.has_email,
        verboseName: 'has an email',
        additionalArguments: 0,
        categoryName: 'Has Email',
        visibility: TEXT_TYPES
    },
    {
        type: Operators.has_error,
        verboseName: 'has an error',
        additionalArguments: 0,
        categoryName: 'Has Error',
        visibility: HIDDEN
    },
    {
        type: Operators.has_value,
        verboseName: 'is not empty',
        additionalArguments: 0,
        categoryName: 'Not Empty'
    },
    {
        type: Operators.has_attachment,
        verboseName: 'has an attachment',
        additionalArguments: 0,
        initialArgument: ARGUMENT_ATTACHMENTS,
        categoryName: 'Has Attachment'
    },
    {
        type: Operators.has_wait_timed_out,
        verboseName: null,
        additionalArguments: 0,
        visibility: HIDDEN
    },
    {
        type: Operators.has_pattern,
        verboseName: 'matches regex',
        additionalArguments: 1
    }
];

export const operatorConfigMap: OperatorMap = operatorConfigList.reduce(
    (map: OperatorMap, operatorConfig: Operator) => {
        map[operatorConfig.type] = operatorConfig;
        return map;
    },
    {}
);

/**
 * Shortcut for constant lookup of operator config in operator configs map
 * @param {string} type - The type of the operator config to return, e.g. 'send_msg'
 * @returns {Object} - The operator config found at operatorConfigs[type] or -1
 */
export const getOperatorConfig = (type: Operators): Operator => operatorConfigMap[type];
