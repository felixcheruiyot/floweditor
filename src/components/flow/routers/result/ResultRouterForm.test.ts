import { RouterFormProps } from '~/components/flow/props';
import ResultRouterForm from '~/components/flow/routers/result/ResultRouterForm';
import { Types } from '~/config/interfaces';
import { AssetType } from '~/store/flowContext';
import { composeComponentTestUtils, mock } from '~/testUtils';
import { createRenderNode, getRouterFormProps } from '~/testUtils/assetCreators';
import * as utils from '~/utils';

const { setup } = composeComponentTestUtils<RouterFormProps>(
    ResultRouterForm,
    getRouterFormProps(
        createRenderNode({
            actions: [],
            exits: [],
            ui: {
                position: { left: 0, top: 0 },
                type: Types.split_by_run_result,
                config: {
                    router: {
                        operandAsset: {
                            id: 'favorite_color',
                            name: 'Favorite Color',
                            type: AssetType.Result
                        }
                    }
                }
            }
        })
    )
);

mock(utils, 'createUUID', utils.seededUUIDs());

describe(ResultRouterForm.name, () => {
    it('should render', () => {
        const { wrapper } = setup(true);
        expect(wrapper).toMatchSnapshot();
    });
});
