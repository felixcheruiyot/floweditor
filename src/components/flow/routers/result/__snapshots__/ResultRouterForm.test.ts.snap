// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResultRouterForm should render 1`] = `
<Dialog
  buttons={
    Object {
      "primary": Object {
        "name": "Ok",
        "onClick": [Function],
      },
      "secondary": Object {
        "name": "Cancel",
        "onClick": [Function],
      },
    }
  }
  headerClass="split_by_run_result"
  tabs={
    Array [
      Object {
        "body": <div
          className="shouldDelimit"
        >
          <CheckboxElement
            checked={false}
            description="Evaluate your rules against a delimited part of your result"
            name="Delimit"
            onChange={[Function]}
            title="Delimit Result"
          />
        </div>,
        "checked": false,
        "name": "Advanced",
      },
    ]
  }
  title="Split by Flow Result"
>
  <TypeList
    __className=""
    initialType={
      Object {
        "aliases": Array [
          "split_by_run_result_delimited",
        ],
        "description": "Split by a result in the flow",
        "form": undefined,
        "localization": [Function],
        "localizeableKeys": Array [
          "exits",
          "cases",
        ],
        "name": "Split by Flow Result",
        "type": "split_by_run_result",
      }
    }
    onChange={[MockFunction]}
  />
  <div
    className="leadIn"
  >
    If the flow result
  </div>
  <div
    className="resultSelect"
  >
    <AssetSelector
      assets={
        Object {
          "items": Object {},
          "type": "result",
        }
      }
      entry={
        Object {
          "value": null,
        }
      }
      name="Flow Result"
      onChange={[Function]}
      placeholder="Select Result"
      searchable={false}
      styles={
        Object {
          "control": [Function],
          "indicatorsContainer": [Function],
          "input": [Function],
          "menuList": [Function],
          "option": [Function],
          "singleValue": [Function],
          "valueContainer": [Function],
        }
      }
    />
  </div>
  <CaseList
    cases={Array []}
    data-spec="cases"
    onCasesUpdated={[Function]}
  />
  <OptionalTextInput
    helpText="By naming the result, you can reference it later using @run.results.whatever_the_name_is"
    name="Result Name"
    onChange={[Function]}
    toggleText="Save as.."
    value={
      Object {
        "value": "",
      }
    }
  />
</Dialog>
`;
